zebra-network/src/peer/client.rs

```diff
@@ -253,31 +261,53 @@ impl<T: std::fmt::Debug> Drop for MustUseOneshotSender<T> {
 impl Client {
     /// Check if this connection's heartbeat task has exited.
     fn check_heartbeat(&mut self, cx: &mut Context<'_>) -> Result<(), SharedPeerError> {
-        if let Poll::Ready(()) = self
+        let is_canceled = self
             .shutdown_tx
             .as_mut()
             .expect("only taken on drop")
             .poll_canceled(cx)
-        {
-            // Make sure there is an error in the slot
-            let heartbeat_error: SharedPeerError = PeerError::HeartbeatTaskExited.into();
-            let original_error = self.error_slot.try_update_error(heartbeat_error.clone());
-            debug!(
-                ?original_error,
-                latest_error = ?heartbeat_error,
-                "client heartbeat task exited"
-            );
+            .is_ready();
 
-            if let Err(AlreadyErrored { original_error }) = original_error {
-                Err(original_error)
-            } else {
-                Err(heartbeat_error)
-            }
+        let has_stopped = self.heartbeat_task.poll_unpin(cx).is_ready();
```

Comment:
Note: this code ignores the `JoinHandle` return type, using `matches!(_, Poll::Ready(()))` would check it at compile time.

The return type might change with future Zebra changes.

Reply:
`JoinHandle` is a `Future<Output = Result<_, JoinError>>`, where `_` is the output type of the task's future (`()` in our case). I'd rather keep it as is, because the current semantics is if the background task has stopped _for any reason_, `Client::poll_ready` should error.

Reply:
I moved the other cleanup tasks into #3263.

That leaves the panics and type-safety to do in this PR. (Sorry, I didn't get a chance to fix them today.)